src/deploy/functions/params.ts

```diff
@@ -0,0 +1,342 @@
+ *   - if the default CEL expression is not resolvable--it depends on a param whose value is not yet known--we throw an error
+ *   - yes, this means that the same set of params may or may not throw depending on the order the SDK provides them to us in
+ *   - after prompting, the resolved value of the param is written to the most specific .env file available
+ */
+export async function resolveParams(
+  params: Param[],
+  projectId: string,
+  userEnvs: Record<string, ParamValue>
+): Promise<Record<string, ParamValue>> {
+  const paramValues: Record<string, ParamValue> = {};
+
+  for (const param of params) {
+    if (userEnvs.hasOwnProperty(param.param)) {
+      if (canSatisfyParam(param, userEnvs[param.param])) {
+        paramValues[param.param] = userEnvs[param.param];
+        continue;
+      } else {
+        throw new FirebaseError(
+          "Parameter " +
+            param.param +
+            " resolved to value from dotenv files " +
+            userEnvs[param.param] +
+            " of wrong type"
+        );
+      }
+    }
+
+    let paramDefault: ParamValue | undefined = param.default;
+    if (paramDefault && isCEL(paramDefault)) {
+      paramDefault = resolveDefaultCEL(param.type || "string", paramDefault, paramValues);
+    }
+    if (paramDefault && !canSatisfyParam(param, paramDefault)) {
+      throw new FirebaseError(
+        "Parameter " + param.param + " has default value " + paramDefault + " of wrong type"
+      );
+    }
+    paramValues[param.param] = await promptParam(param, paramDefault);
+  }
+
+  return paramValues;
+}
+
+/**
+ * Returns the resolved value of a user-defined Functions parameter.
+ * Functions params are defined by the output of the Functions SDK, but their value is not set until deploy-time.
+ *
+ * For most param types, we check the contents of the dotenv files first for a matching key, then interactively prompt the user.
+ * When the CLI is running in non-interactive mode or with the --force argument, it is an error for a param to be undefined in dotenvs.
+ */
+async function promptParam(param: Param, resolvedDefault?: ParamValue): Promise<ParamValue> {
+  if (param.type === "string") {
+    return promptStringParam(param, resolvedDefault as string);
+  } else if (param.type === "int") {
+    return promptIntParam(param, resolvedDefault as number);
+  } else {
+    assertExhaustive(param);
+  }
+}
+
+async function promptStringParam(param: StringParam, resolvedDefault?: string): Promise<string> {
+  if (!param.input) {
+    const defaultToText: TextInput<string> = { text: {} };
+    param.input = defaultToText;
+  }
+
+  switch (param.input.type) {
+    case "select":
+      throw new FirebaseError(
+        "Build specified string parameter " + param.param + " with unsupported input type 'select'"
+      );
+    case "text":
+    default:
+      let prompt = `Enter a value for ${param.param}`;
+      if (param.label) {
+        prompt = `${prompt} (${param.label})`;
+      }
```

Comment:
I _suspect_ the better developer experience is to say `Enter a value for ${param.label || param.param}` (making the env var name meaningless when the user supplied a human readable label). If you think that's not obvious we should bring up the point in the CF3 chat room.

Reply:
Done. Agreed.